oresat_configs/scripts/gen_kaitai.py

```diff
@@ -0,0 +1,226 @@
+                        "id": "rpt_instance",
+                        "type": "repeaters",
+                        "repeat": "until",
+                        "repeat-until": "until",
+                        "doc": "Repeat until no repeater flag is set!",
+                    }
+                ]
+            },
+            "repeaters": {
+                "seq": [
+                    {
+                        "id": "rpt_callsign_raw",
+                        "type": "callsign_raw",
+                    },
+                    {
+                        "id": "rpt_ssid_raw",
+                        "type": "ssid_mask",
+                    },
+                ]
+            },
+            "callsign_raw": {
+                "seq": [
+                    {
+                        "id": "callsign_ror",
+                        "process": "repeaters",
+                        "size": 6,
+                        "type": "callsign",
+                    }
+                ]
+            },
+            "callsign": {
+                "seq": [
+                    {
+                        "id": "callsign",
+                        "type": "str",
+                        "encoding": "ASCII",
+                        "size": 6,
+                        "valid": {"any-of": ['"KJ7SAT"', '"SPACE "']},
+                    }
+                ]
+            },
+            "ssid_mask": {
+                "seq": [
+                    {
+                        "id": "ssid_mask",
+                        "type": "u1",
+                    }
+                ],
+                "instances": {"ssid": {"value": "(ssid_mask & 0x0f) >> 1"}},
+            },
+            "i_frame": {
+                "seq": [
+                    {
+                        "id": "pid",
+                        "type": "u1",
+                    },
+                    {"id": "ax25_info", "type": "ax25_info_data", "size-eos": True},
+                ]
+            },
+            "ui_frame": {
+                "seq": [
+                    {
+                        "id": "pid",
+                        "type": "u1",
+                    },
+                    {"id": "ax25_info", "type": "ax25_info_data", "size-eos": True},
+                ]
+            },
+            "ax25_info_data": {"seq": []},
+        },
+    }
+
+    # Append field types for each field
+    for obj in config.beacon_def:
+        new_var = {
+            "id": obj.name,
+            "type": CANOPEN_TO_KAITAI_DT[obj.data_type],
+            "doc": obj.description,
+        }
+        if new_var["type"] == "str":
+            new_var["encoding"] = "ASCII"
+            if obj.access_type == "const":
+                new_var["size"] = len(obj.default)
+
+        cast(
+            Any, cast(Any, cast(Any, kaitai_data.get("types")).get("ax25_info_data")).get("seq")
+        ).append(
+            new_var
+        )  # Same as: `kaitai_data["types"]["ax25_info_data"]["seq"].append(new_var)`D
```

Comment:
Amusing, but there's easier ways of telling the type checker to go away. The error being thrown here is in essence telling us that mypy didn't resolve the type of `kaitai_data` completely. Understandable, since heavily nested dicts can be a pain. We fix this by giving a type hint to `kaitai_data`. To just tell mypy to go away is `kaitai_data: Any = { ...`, but we can do at least a little better: `kaitai_data: dict[str, Any] = { ...`. More complete types would require a different approach to be not obnoxious.

Reply:
> Trying to do the sane way of `dict[str, dict[str, Any]]` didn't seem to work last I tried it, FWIW.

It's because `kaitai_data` is really a `dict[str, Union[str, list, dict]]` (see keys `doc` and `seq`). Then because the value is a union, when you access any key you need to ensure the returned value is the type you were expecting with `isinstance()`.

If you did want to try two layers of typing you might use [`TypedDict`](https://docs.python.org/3/library/typing.html#typing.TypedDict).

Reply:
Updated